scala/src/main/scala/Main.scala

```diff
@@ -23,35 +23,37 @@ def randomCollisionShape(using g: Random): CollisionShape =
         Vector3.randomIn(AxisAlignedBox.unit),
         Vector3.randomIn(AxisAlignedBox.unit))
     case n =>
-      throw new RuntimeException("unreachable")
-
-@main def Main =
-  given Random = scala.util.Random()
+      throw new RuntimeException("unreachable: " + n)
 
+def run(size: Int)(using Random): Int =
   // Create an empty world.
   var world = Scene()
 
   // Add random objects to the world.
-  for _ <- 0 until 1000 do
+  for _ <- 0 until size do
     val n = Node()
     n.translation = Vector3.randomIn(worldBounds)
     n.shape = Some(randomCollisionShape)
     world = world.adding(n)
 
   // Shoot rays at all objects from the origin to find those that are (partially) occluded.
-  var nodes = world.nodes
+  var nodes = Set.from(world.nodes)
   var occluded = List[Node]()
 
-  while !nodes.isEmpty do
-    val n = nodes.head
-    nodes = nodes.drop(1)
-
+  for n <- world.nodes if nodes(n) do
+    //println(s"Node: ${n.translation}")
     val s = n.shape.get
     val r = Ray(Vector3.zero, (world.translation(n) + s.centroid).normalized)
+    //println(s"Ray: $r")
     val collisions = world.shoot(r, -1, false)
     for (m, _) <- collisions do
       n.collisionMask = 0
-      world = world.removing(m)
```

Comment:
Shouldn't we remove `m` from `nodes`, as in the Swift version?

Reply:
I think the swift version is maintaining a worklist whereas we're just iterating over the set here. That should be completely equivalent because we're removing one node at each iteration. If anything it is at Scala's advantage because set removal is not as cheap as advancing an iterator.